go/vt/binlog/binlogplayer/binlog_player.go

```diff
@@ -703,6 +709,49 @@ func ReadVReplicationStatus(index uint32) string {
 	return fmt.Sprintf("select pos, state, message from _vt.vreplication where id=%v", index)
 }
 
+// MysqlUncompress will uncompress a binary string in the format stored by mysql's compress() function
+// The first four bytes represent the size of the original string passed to compress()
+// Remaining part is the compressed string using zlib, which we uncompress here using golang's zlib library
+func MysqlUncompress(input string) []byte {
+	// consistency check
+	inputBytes := []byte(input)
+	if len(inputBytes) < 5 {
+		return nil
+	}
+
+	// determine length
+	dataLength := uint32(inputBytes[0]) + uint32(inputBytes[1])<<8 + uint32(inputBytes[2])<<16 + uint32(inputBytes[3])<<24
+	dataLengthBytes := make([]byte, 4)
+	binary.LittleEndian.PutUint32(dataLengthBytes, dataLength)
+	dataLength = binary.LittleEndian.Uint32(dataLengthBytes)
+
+	// uncompress using zlib
+	inputData := inputBytes[4:]
```

Comment:
I got curious why this bit math was required and not provided by `compress/zlib` so I started to play around with this.

From my read and brief experimentation to write some benchmarks, that golang's zlib library expects this header to be present, so chopping off the first four bytes will always result in a `zlib.ErrHeader` result from `zlib.NewReader`: https://play.golang.org/p/p6XWE12Um89

So, I commented out the header truncation, and was able to exercise the rest of the decompression code as written, but it would still fail on the length check (line 740), which I haven't looked into yet. In any case, the results between doing this vs just calling into `NewReader` and then checking if the error is `zlib.ErrHeader` (which would indicate the data was not compressed with zlib) seem negligible. [Benchmark repo](https://github.com/ajm188/zlib-example/blob/main/benchmarks/uncompress_bench_test.go#L11-L29).

**Note**: I did not check this against gtids compressed by mysql yet, so we should verify if the behavior of mysql's `compress` is actually different from golang's zlib compression. From [the docs](https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_compress), it _seems_ (but I am very unfamiliar, so happy to be told I'm wrong here), that mysql doesn't guarantee to be using zlib (due to "a compression library _such as_ zlib" vs "this compression is always done with specifically zlib")

Reply:
The four bytes are additionally added by MySQL. They are not related the header of zlib-generated data.  Some references here:
```
https://bugs.mysql.com/bug.php?id=79400
https://www.bennadel.com/blog/3152-running-mysql-compress-and-uncompress-compatible-methods-in-coldfusion.htm
```
https://dev.mysql.com/doc/refman/8.0/en/innodb-compression-internals.html says zlib is used. Maybe there is a way to use a different library while compiling MySQL but I haven't seen references to anyone using a different library. Also I saw references to tools and drivers that have assumed zlib. 

So if it is possible to link in other compression libraries and someone does it, we will have to update this implementation.

Reply:
Thanks for the clarification!! This makes sense to me now